### PR TITLE
[ENH] Implement scanners for performance enhancement on structured data

### PR DESCRIPTION
**Pull request by zzzzwj**
This PR implements the function on extracting all local patches from a image dataset, which is similar to ```sklearn.feature_extraction.image.extract_patches_2d``` but with more options, such as the kernel size, stride, padding size.

It is used in the original paper and designed to improve the performance of deep forest on image datasets. In addition, the scanner can also be used to pre-process the dataset and improve the performance of other classifiers.

After transformations, the training dataset will be in the shape ```(n_samples * n_patches_per_sample, kernel_size*kernel_size)```. Notice that since ```n_samples * n_patches_per_sample``` is typically very large, the data type of feature values should be cast into ```np.unit8``` after using the scanner.

Steps:
- [x] Add the PyTorch backend and GPU support
- [x] Add the Numpy backend
- [x] Add unit tests on the image scanner (Input: a toy image; Test: All patches should be exactly what we want)
- [ ] Add the documentation
- [ ] Verify the improvements after using the image scanner

### CODE REVIEW

File: deepforest/scanner.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+"""
+Implement methods on extracting structured data such as images or sequences.
+"""
+import torch
+import torch.nn as nn
```
Comment (xuyxu) [inline deepforest/scanner.py:5]: This will make `torch` a hard dependency for Deep Forest. We'd better use the `__import__` statement. You can refer to [here](https://github.com/LAMDA-NJU/Deep-Forest/blob/master/deepforest/cascade.py#55) for an example.

File: deepforest/scanner.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        patches = patches.view(n, c, -1, patch_sz, patch_sz)
+        patches = patches.permute(0, 2, 1, 3, 4)
+        patches = patches.view(-1, c, patch_sz, patch_sz)
+        return patches
+
+    def _numpy_transform(self, X):
+        pass
+
+    def fit_transform(self, X):
+        self._validate_parameters(X.shape)
```
Comment (xuyxu) [inline deepforest/scanner.py:47]: The input `X` is a numpy array of any data type, we need to wrap it using `TensorDataset` before passing it into `_torch_transform`. Therefore, in the IF body of `self.backend == torch`, we should first conduct a transformation from Numpy array to TensorDataset.

File: deepforest/scanner.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        self.backend = backend
+
+    def _validate_parameters(self, image_size):
+        pass
+
+    def _torch_transform(self, X):
+        """the shape of X is n, c, h, w"""
+        err_msg = "_torch_transform only accepts torch.Tensor as input, \
+                  but got {}"
+        assert isinstance(X, torch.Tensor), err_msg.format(type(X))
```
Comment (xuyxu) [inline deepforest/scanner.py:23]: See the comment on `fit_transform`

File: deepforest/scanner.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        if X.dtype != torch.uint8:
+            msg = "Warning! ImageScanner will force to transform data \
+                   type of {} to uint8."
+            print(msg.format(X.dtype))
+            X = X.byte()
```
Comment (xuyxu) [inline deepforest/scanner.py:28]: The input to `_torch_transform` is a TensorDataset, which wraps the original dataset in the form of Numpy array. The output of `_torch_transform` is a Numpy array of the size (n_samples * n_patches_per_sample, kernel_size * kernel_size), and the data type is `np.uint8`.

**Review by xuyxu — COMMENTED** [review]
Hi @zzzzwj, thanks for the PR. I have made some clarifications on the function `_torch_transform`, and feel free to tell me if you are confused about something.

File: deepforest/scanner.py (hunk 3 above)
Comment (zzzzwj) [reply deepforest/scanner.py:23]: Got it!

**Comment by xuyxu** [comment]
Hi @zzzzwj, codes and unit tests on the pytorch backend look great! In addition, I have added some docstrings on the functions, and changed the place of `__import__` to avoid the `ModuleNotFoundError` of `torch` when we are using the `numpy` backend. Let me know if you are fine with these modifications ;-)

Besides, could you further complete two tasks listed below?
* **Add GPU support on the pytorch backend.**
  * I think it is fine to add another parameter: `device` to `__init__`, and the default value is `"cpu"`. After then, we can follow the rountine in PyTorch: use a CPU / GPU device in the function `_torch_transform`, and call `X = X.to(device)`. Notice that the final result from `_torch_transform` should be on the side of CPU.
* **Runtime results on MNIST and CIFAR-10 dataset when we are using the CPU and GPU to extract image patches.**

Please feel free to ask me if you have any problem, thanks!

**Comment by zzzzwj** [comment]
@xuyxu Good job. Now this version is really nice to me. Thanks for your work! And I'll submit the task of GPU support and runtime results in few days.

**Comment by zzzzwj** [comment]
The runtime results of CPU & GPU mode in pytorch backend is shown below: (kernel_size=5, stride=1, padding=0)
```
 Dataset | Device |  Elapsed Time   |   Size of extracted patches
  MNIST  |  CPU   |  0.575(±0.03)s  |          823.97MB
  MNIST  |  GPU   |  0.344(±0.02)s  |          823.97MB
 CIFAR10 |  CPU   |  2.729(±0.10)s  |          2803.80MB
 CIFAR10 |  GPU   |  1.179(±0.07)s  |          2803.80MB
```
The runtime experiments are run on a machine with Intel  i7-8700K & one Nvidia 1080Ti GPU. And the values are recorded by average results of 30 tests.

**Comment by xuyxu** [comment]
Thanks for reporting the results! I will fix the code quality issue.

**Comment by xuyxu** [comment]
Hi @zzzzwj, this feature request is more complicated than I expected. I have moved the image scanner into a sub-module named `preprocessing`, implementations on the scanner for deep forest is still a on-going work.

Meanwhile, could you further improve unit tests on the image scanner? I think it would be better if the toy data used for testing can have more than one samples. Thanks ;-)

**Comment by zzzzwj** [comment]
@xuyxu OK, I will.

**Review by xuyxu — COMMENTED** [review]
LGTM.

**Comment by xuyxu** [comment]
Thanks for the following-up code @zzzzwj!

I have created a new branch `scanning`, since there is still a long way to go before completing this functionality. Could you contribute your PR into that branch instead of the master branch? You may need to open another PR with exactly the same codes. Thanks !
